README
Installation of the Yubikey Personalization package
===================================================

Yubikey Personalization
-----------------------

The YubiKey Personalization package contains a library and command
line tool used to personalize (i.e., set a AES key) YubiKeys.

Documentation
-------------

The complete reference manual on the YubiKey is required reading if
you want to understand the entire picture and what each parameter
does.  Download it from http://www.yubico.com/

Dependencies
------------

Getting and installing dependencies depends on your operating systems,
we give example for some flavours.  If you know how to install
dependencies on other systems, let us know.  Debian hints should apply
to Debian derivatives as well, including Ubuntu.

Yubico-c is needed, see: https://developers.yubico.com/yubico-c/

  Debian:           apt-get install libyubikey-dev
  Fedora:           dnf install libyubikey-devel

Pkg-config simplify finding other dependencies, see:
http://www.freedesktop.org/wiki/Software/pkg-config

  Debian:           apt-get install pkg-config

Yubikey-personalization depends on libusb or libusb-1, so you will
have to get it.  We recommend using libusb-1.

  Debian libusb-1:  apt-get install libusb-1.0-0-dev
  Debian libusb:    apt-get install libusb-dev
  Fedora:           dnf install libusb-devel

The JSON library is an optional dependency, see:
https://github.com/json-c/json-c/wiki

  Debian:           apt-get install libjson-c-dev

You need json-c version 0.10 or later to get pretty printing of JSON
output.  This project will build with version 0.9 too, but will not
pretty print the JSON output.

License
-------

The project is licensed under a BSD license.  See the file COPYING for
exact wording.  For any copyright year range specified as YYYY-ZZZZ in
this package note that the range specifies every single year in that
closed interval.

Building from Git
-----------------

Skip to the next section if you are using an official packaged
version.

You may check out the sources using Git with the following command:

-----------
  git clone https://github.com/Yubico/yubikey-personalization.git
-----------

This will create a directory 'yubikey-personalization'.  Enter the directory:

-----------
  cd yubikey-personalization
-----------


When building from source Yubikey-personalization depends on link:http://asciidoc.org/INSTALL.html[asciidoc], xsltproc and DocBook to build its manpage.

-----------
  Debian:           apt-get install asciidoc xsltproc docbook-xsl
-----------

Autoconf, automake and libtool must be installed.

Generate the build system using:

-----------
  autoreconf --install
-----------

Building
--------

The build system uses Autoconf, to set up the build system run:

-----------
  ./configure
-----------

Then build the code, run the self-test and install the binaries:

-----------
  make check install
-----------

Using
-----

WARNING: By using this tool you will destroy the AES key in your
YubiKey.  This prevents it from being useful against Yubico's
validation server.  It is possible to upload a new AES key to Yubico,
using a random YubiKey prefix, to restore it.  But it is not possible
to get back your old yubikey prefix if you decide to re-program your
YubiKey.

IMPORTANT: When running any of the utils that need to access the YubiKey 
you will either need to run as root, or you will have to have made sure 
that the current user has permission to access the device. These 
permissions can be set up by copying the udev rules files 
(https://github.com/Yubico/yubikey-personalization/blob/master/69-yubikey.rules[69-yubikey.rules] 
and https://github.com/Yubico/yubikey-personalization/blob/master/70-yubikey.rules[70-yubikey.rules]) to /etc/udev/rules.d/

With that out of the way, here is how you would program a YubiKey with
an all-zero AES key and a dummy prefix:

-----------
$ ./ykpersonalize -1 -ofixed=cccccccccccc -a00000000000000000000000000000000
Firmware version 1.3.1 Touch level 9840 Program sequence 10
Configuration data to be written to key configuration 1:

fixed: m:cccccccccccc
uid: h:000000000000
key: h:00000000000000000000000000000000
acc_code: h:000000000000
ticket_flags: APPEND_CR
config_flags:

Commit? (y/n) [n]: y
$ 
-----------

Using the "ykparse" tool from the yubico-c package, you can check that
the OTPs are correct.  For example:

-----------
$ ykparse 00000000000000000000000000000000 ccccccccccccdkrkedgchtlfefghcekefhlifbchijrd
warning: overlong token, ignoring prefix: cccccccccccc
Input:
  token: dkrkedgchtlfefghcekefhlifbchijrd
          29 c9 32 50 6d a4 34 56 03 93 46 a7 41 06 78 c2 
  aeskey: 00000000000000000000000000000000
          00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
Output:
          00 00 00 00 00 00 01 00 53 ea 63 00 6f 9e c4 24 

Struct:
  uid: 00 00 00 00 00 00 
  counter: 1 (0x0001)
  timestamp (low): 59987 (0xea53)
  timestamp (high): 99 (0x63)
  session use: 0 (0x00)
  random: 40559 (0x9e6f)
  crc: 9412 (0x24c4)

Derived:
  cleaned counter: 1 (0x0001)
  modhex uid: cccccccccccc
  triggered by caps lock: no
  crc: F0B8
  crc check: ok
$ 
-----------

To program a YubiKey in static mode, you use the -ostatic-ticket flag
as follows:

-----------
$  ./ykpersonalize -1 -ofixed=cccccccccccc -a00000000000000000000000000000000 -ostatic-ticket
Firmware version 1.3.1 Touch level 9856 Program sequence 11
Configuration data to be written to key configuration 1:

fixed: m:cccccccccccc
uid: h:000000000000
key: h:00000000000000000000000000000000
acc_code: h:000000000000
ticket_flags: APPEND_CR
config_flags: STATIC_TICKET

Commit? (y/n) [n]: y
$ 
-----------

To program a YubiKey in static mode with a strongly looking password
(i.e., also containing numeric and upper case letters), you use the
-ostatic-ticket flag together with -ostrong-pw1 and -ostrong-pw2 (note
YubiKey 2.0 only!) as follows:

-----------
$  ./ykpersonalize -1 -ofixed=cccccccccccc -a00000000000000000000000000000000 -ostatic-ticket -ostrong-pw1 -ostrong-pw2
Firmware version 2.0.0 Touch level 1792 Program sequence 3
Configuration data to be written to key configuration 1:

fixed: m:cccccccccccc
uid: h:000000000000
key: h:00000000000000000000000000000000
acc_code: h:000000000000
ticket_flags: APPEND_CR
config_flags: STATIC_TICKET|STRONG_PW1|STRONG_PW2

Commit? (y/n) [n]: y
$ 
-----------

Alternatively on a YubiKey 2.0, you can program the second configuration, which
defaults to be the static key configuration:

-----------
$  ./ykpersonalize -2 -ofixed=cccccccccccc -a00000000000000000000000000000000
Firmware version 2.0.0 Touch level 1792 Program sequence 3
Configuration data to be written to key configuration 2:

fixed: m:cccccccccccc
uid: h:000000000000
key: h:00000000000000000000000000000000
acc_code: h:000000000000
ticket_flags: APPEND_CR
config_flags: STATIC_TICKET|STRONG_PW1|STRONG_PW2

Commit? (y/n) [n]: y
$ 
-----------

To program a YubiKey with a lock code (to prevent others from easily
reprogramming it), you use the -oaccess= flag as follows:

-----------
$ ./ykpersonalize -1 -ofixed=vvvecdcedvjj -a00000000000000000000000000000000 -oaccess=001100001100
Firmware version 2.0.0 Touch level 1792 Program sequence 3
Configuration data to be written to key configuration 1:

fixed: m:vvvecdcedvjj
uid: h:000000000000
key: h:00000000000000000000000000000000
acc_code: h:001100001100
ticket_flags: APPEND_CR
config_flags:

Commit? (y/n) [n]: y
$
-----------

To re-program a YubiKey that has a lock code set, you use the
-cXXX.. flag as follows:

-----------
$ ./ykpersonalize -1 -c001100001100 -ofixed=vvvecdcedvjj -a00000000000000000000000000000000 -oaccess=001100223300
Firmware version 2.0.0 Touch level 1792 Program sequence 3
Configuration data to be written to key configuration 1:

fixed: m:vvvecdcedvjj
uid: h:000000000000
key: h:00000000000000000000000000000000
acc_code: h:001100223300
ticket_flags: APPEND_CR
config_flags:

Commit? (y/n) [n]: y
$
-----------

To disable the lock code on a YubiKey, program it with a lock code set
to zeros.  For example:

-----------
$ ./ykpersonalize -1 -c001100001133 -ofixed=vvvecdcedvjj -a00000000000000000000000000000003 -oaccess=000000000000
Firmware version 2.0.0 Touch level 1792 Program sequence 7
Configuration data to be written to key configuration 1:

fixed: m:vvvecdcedvjj
uid: h:000000000000
key: h:00000000000000000000000000000000
acc_code: h:000000000000
ticket_flags: APPEND_CR
config_flags:

Commit? (y/n) [n]: y
$
-----------


NEWS
Yubikey-personalize NEWS -- History of user-visible changes.     -*- outline -*-

* Version 1.20.1 (unreleased)

* Version 1.20.0 (released 2019-07-03)

** Add yk_open_key_vid_pid() allowing vid and pid to be specified.

** Documentation fixes.

** Clear potentially sensitive material from buffers.

** Fix potential buffer overwrite.

* Version 1.19.3 (released 2019-02-22)

** Fix capability read.

* Version 1.19.2 (released 2019-02-19)

** Fix test on mac.

** Fix serial read and challenge response.

* Version 1.19.1 (released 2019-02-19)

** Error out on json output with randomSeed.

** Validate more length fields.

** Use correct FormatMessage function on windows.

** Overflow, bounds and error condition checks.

** Try to zero sensitive memory better.

* Version 1.19.0 (released 2018-04-24)

** Add yk_write_device_info().

** Add ykpersonalize cli switch -D for device info.

** Add code for handling personalization interface of major version 5.

* Version 1.18.1 (released 2018-01-16)

** Support reading accesscode and private ID from stdin.

** Parse optional arguments correctly.

** Documentation fixes.

** Fix for ykinfo modhex serial output when it ends with c.

** Treat all firmware versions as supported.

* Version 1.18.0 (released 2017-01-27)

** Let ykchalresp read challenge from a file.

** Add support of working with a numbered key when many connected
Thanks to Thomas Habets <habets@google.com>

** Documentation clarifications.

** Fixup argument parsing of flags with optional arguments on BSD platforms.

** Fix a file descriptor leak on windows.

* Version 1.17.3 (released 2015-12-28)

** Dont read to much if we don't find a key.

** Text updates to make options clearer.

** Correct logic for question when mode switching to non-otp mode.

** Add 4.3 as supported firmware.

* Version 1.17.2 (released 2015-09-22)

** Let _yk_write() return an error if yk_wait_for_key_status() fails.

** Fix a mistake in help, fixed is up to 16 bytes, 32 characters.

** Add 4.2 as supported firmware.

* Version 1.17.1 (released 2015-04-01)

** Fixup of 1.17.0

* Version 1.17.0 (released 2015-04-01)

** add yk_get_capabilities() to fetch capabilities.

** add -c to ykinfo to fetch capabilities.

** whitelist firmware 4.1.x

* Version 1.16.4 (released 2015-03-23)

** change the tool to accept autoeject time as a short instead of a byte

* Version 1.16.3 (released 2015-03-10)

** whitelist YubiKey version 3.4.x

** only try to set libusb configuration if it's unset on the device

* Version 1.16.2 (released 2014-11-28)

** ykinfo: fix modhex printout when serial is an odd number of hex digits.

** whitelist yubikey version 4.0.x

** try to open more PIDs and add for udev.

* Version 1.16.1 (released 2014-10-30)

** Fix udev rules so they contain four digits.

** Only try to detach the kernel driver if it's attached.
For libusb-1.0

** Let import config report errors properly.

* Version 1.16.0 (released 2014-09-26)

** Add yk_get_key_vid_pid() to get the vendor and product id of a key.

** Add flags for ykinfo to print vendor and product id.

** Fix a bug in the osx backend where it would return an error opening
a composite device with two hid interfaces.

* Version 1.15.3 (released 2014-09-04)

** Fix URLs for opensource.y.com -> developers.y.com move.

** Whitelist firmware version 3.3 and detect new PIDs.

* Version 1.15.2 (released 2014-07-30)

** Whitelist firmware version 2.5

** Read key when importing configuration.

** Fix formatting error in information about what is written to key.

** Check return codes when doinf NDEF writes.

* Version 1.15.1 (released 2014-03-12)

** Add -6 and -8 flags to ykchalresp to output the response as a truncated
OATH code. Thanks to joernchen of Phenoelit <joernchen@phenoelit.de>

** Add -t to ykchalresp to use current time / 30 as challenge in TOTP
fashion. Thanks to joernchen of Phenoelit <joernchen@phenoelit.de>

** Fix for a crash in the libusb-0.1 backend.
Thanks to Antti Keränen <detegr@gmail.com>

* Version 1.15.0 (released 2014-01-10)

** add new Neo version 3.2 as supported.

** let ykp_AES_key_from_password() return error if it can't get a random
salt.

** let -a without option mean ask for raw key on STDIN.

** let no -a given mean generate a random key

** add ykp_AES_key_from_raw() and ykp_HMAC_key_from_raw() that sets the
key from a byte array.

** remove -osalt= and possibility to generate key from password.

** add YK_ENODATA that's returned when no data is returned from
the device and no other underlying error is set.

* Version 1.14.2 (released 2013-12-30)

** let -m take challenge-response timeout and autoeject timeout
supplied with : beteen, so mode:cr-timeout:autoeject-timeout

** make sure to always wait 1 millisecond between writes to the key
newer platforms triggered a race where writes to the key could
happen in a very short (about 0.4 milliseconds) timespan and would
not be processed.

** minor fixes for the printing format.

* Version 1.14.1 (released 2013-09-16)

** Fix breakage with latest automake.

* Version 1.14.0 (released 2013-07-04)

** Fixups of import/export.
Add targetConfig to show in which slot a configuration is intended.
Possible memory leaks on error conditions.

** Add -d switch to ykpersonalize for dry-run.

** Add ykp_clear_config() for clearing configuration flags.

** Add getter functions for all configuration flags.

** Add -V to all tools to output version.

** Add ykp_get_acccode_type() and ykp_set_acccode_type()
Only to do with export, showing where the access code came from
in the ycfg.

** Add -1 and -2 options to ykinfo to show programming state.

* Version 1.13.0 (released 2013-04-22)

** Added import/export of ycfg-json format.
Invoke with -fjson to -s or -i
Add exported functions ykp_export_config() and ykp_import_config()

** Fixup output of flags when using ykp_write_config()

** Add binary builds for mac.

** Minor cleanups noticed during debian packaging.

* Version 1.12.0 (released 2013-03-14)


sha1.c
/**************************** sha1.c ****************************/
/******************** See RFC 4634 for details ******************/
/*
 *  Description:
 *      This file implements the Secure Hash Signature Standard
 *      algorithms as defined in the National Institute of Standards
 *      and Technology Federal Information Processing Standards
 *      Publication (FIPS PUB) 180-1 published on April 17, 1995, 180-2
 *      published on August 1, 2002, and the FIPS PUB 180-2 Change
 *      Notice published on February 28, 2004.
 *
 *      A combined document showing all algorithms is available at
 *              http://csrc.nist.gov/publications/fips/
 *              fips180-2/fips180-2withchangenotice.pdf
 *
 *      The SHA-1 algorithm produces a 160-bit message digest for a
 *      given data stream.  It should take about 2**n steps to find a
 *      message with the same digest as a given message and
 *      2**(n/2) to find any two messages with the same digest,
 *      when n is the digest size in bits.  Therefore, this
 *      algorithm can serve as a means of providing a
 *      "fingerprint" for a message.
 *
 *  Portability Issues:
 *      SHA-1 is defined in terms of 32-bit "words".  This code
 *      uses <stdint.h> (included via "sha.h") to define 32 and 8
 *      bit unsigned integer types.  If your C compiler does not
 *      support 32 bit unsigned integers, this code is not
 *      appropriate.
 *
 *  Caveats:
 *      SHA-1 is designed to work with messages less than 2^64 bits
 *      long. This implementation uses SHA1Input() to hash the bits
 *      that are a multiple of the size of an 8-bit character, and then
 *      uses SHA1FinalBits() to hash the final few bits of the input.
 */

#include "sha.h"
#include "sha-private.h"

/*
 *  Define the SHA1 circular left shift macro
 */
#define SHA1_ROTL(bits,word) \
                (((word) << (bits)) | ((word) >> (32-(bits))))

/*
 * add "length" to the length
 */
static uint32_t addTemp;
#define SHA1AddLength(context, length)                     \
    (addTemp = (context)->Length_Low,                      \
     (context)->Corrupted =                                \
        (((context)->Length_Low += (length)) < addTemp) && \
        (++(context)->Length_High == 0) ? 1 : 0)

/* Local Function Prototypes */
static void SHA1Finalize(SHA1Context *context, uint8_t Pad_Byte);
static void SHA1PadMessage(SHA1Context *, uint8_t Pad_Byte);
static void SHA1ProcessMessageBlock(SHA1Context *);

/*
 *  SHA1Reset
 *
 *  Description:
 *      This function will initialize the SHA1Context in preparation
 *      for computing a new SHA1 message digest.
 *
 *  Parameters:
 *      context: [in/out]
 *          The context to reset.
 *
 *  Returns:
 *      sha Error Code.
 *
 */
int SHA1Reset(SHA1Context *context)
{
    if (!context)
        return shaNull;

    context->Length_Low             = 0;
    context->Length_High            = 0;
    context->Message_Block_Index    = 0;

    /* Initial Hash Values: FIPS-180-2 section 5.3.1 */
    context->Intermediate_Hash[0]   = 0x67452301;
    context->Intermediate_Hash[1]   = 0xEFCDAB89;
    context->Intermediate_Hash[2]   = 0x98BADCFE;
    context->Intermediate_Hash[3]   = 0x10325476;
    context->Intermediate_Hash[4]   = 0xC3D2E1F0;

    context->Computed   = 0;
    context->Corrupted  = 0;

    return shaSuccess;
}

/*
 *  SHA1Input
 *
 *  Description:
 *      This function accepts an array of octets as the next portion
 *      of the message.
 *
 *  Parameters:
 *      context: [in/out]
 *          The SHA context to update
 *      message_array: [in]
 *          An array of characters representing the next portion of
 *          the message.
 *      length: [in]
 *          The length of the message in message_array
 *
 *  Returns:
 *      sha Error Code.
 *
 */
int SHA1Input(SHA1Context *context,
    const uint8_t *message_array, unsigned length)
{
  if (!length)
    return shaSuccess;

  if (!context || !message_array)
    return shaNull;

  if (context->Computed) {
    context->Corrupted = shaStateError;
    return shaStateError;
  }

  if (context->Corrupted)
     return context->Corrupted;

  while (length-- && !context->Corrupted) {
    context->Message_Block[context->Message_Block_Index++] =
      (*message_array & 0xFF);

    if (!SHA1AddLength(context, 8) &&
      (context->Message_Block_Index == SHA1_Message_Block_Size))
      SHA1ProcessMessageBlock(context);

    message_array++;
  }

  return shaSuccess;
}

/*
 * SHA1FinalBits
 *
 * Description:
 *   This function will add in any final bits of the message.
 *
 * Parameters:
 *   context: [in/out]
 *     The SHA context to update
 *   message_bits: [in]
 *     The final bits of the message, in the upper portion of the
 *     byte. (Use 0b###00000 instead of 0b00000### to input the
 *     three bits ###.)
 *   length: [in]
 *     The number of bits in message_bits, between 1 and 7.
 *
 * Returns:
 *   sha Error Code.
 */
int SHA1FinalBits(SHA1Context *context, const uint8_t message_bits,
    unsigned int length)
{
  uint8_t masks[8] = {
      /* 0 0b00000000 */ 0x00, /* 1 0b10000000 */ 0x80,
      /* 2 0b11000000 */ 0xC0, /* 3 0b11100000 */ 0xE0,
      /* 4 0b11110000 */ 0xF0, /* 5 0b11111000 */ 0xF8,
      /* 6 0b11111100 */ 0xFC, /* 7 0b11111110 */ 0xFE
  };
  uint8_t markbit[8] = {
      /* 0 0b10000000 */ 0x80, /* 1 0b01000000 */ 0x40,
      /* 2 0b00100000 */ 0x20, /* 3 0b00010000 */ 0x10,
      /* 4 0b00001000 */ 0x08, /* 5 0b00000100 */ 0x04,
      /* 6 0b00000010 */ 0x02, /* 7 0b00000001 */ 0x01
  };

  if (!length)
    return shaSuccess;

  if (!context)
    return shaNull;

  if (context->Computed || (length >= 8) || (length == 0)) {
    context->Corrupted = shaStateError;
    return shaStateError;
  }

  if (context->Corrupted)
     return context->Corrupted;

  SHA1AddLength(context, length);
  SHA1Finalize(context,
    (uint8_t) ((message_bits & masks[length]) | markbit[length]));

  return shaSuccess;
}

/*
 * SHA1Result
 *
 * Description:
 *   This function will return the 160-bit message digest into the
 *   Message_Digest array provided by the caller.
 *   NOTE: The first octet of hash is stored in the 0th element,
 *      the last octet of hash in the 19th element.
 *
 * Parameters:
 *   context: [in/out]
 *     The context to use to calculate the SHA-1 hash.
 *   Message_Digest: [out]
 *     Where the digest is returned.
 *
 * Returns:
 *   sha Error Code.
 *
 */
int SHA1Result(SHA1Context *context,
    uint8_t Message_Digest[SHA1HashSize])
{
  int i;

  if (!context || !Message_Digest)
    return shaNull;

  if (context->Corrupted)
    return context->Corrupted;

  if (!context->Computed)
    SHA1Finalize(context, 0x80);

  for (i = 0; i < SHA1HashSize; ++i)
    Message_Digest[i] = (uint8_t) (context->Intermediate_Hash[i>>2]
              >> 8 * ( 3 - ( i & 0x03 ) ));

  return shaSuccess;
}

/*
 * SHA1Finalize
 *
 * Description:
 *   This helper function finishes off the digest calculations.
 *
 * Parameters:
 *   context: [in/out]
 *     The SHA context to update
 *   Pad_Byte: [in]
 *     The last byte to add to the digest before the 0-padding
 *     and length. This will contain the last bits of the message
 *     followed by another single bit. If the message was an
 *     exact multiple of 8-bits long, Pad_Byte will be 0x80.
 *
 * Returns:
 *   sha Error Code.
 *
 */
static void SHA1Finalize(SHA1Context *context, uint8_t Pad_Byte)
{
  int i;
  SHA1PadMessage(context, Pad_Byte);
  /* message may be sensitive, clear it out */
  for (i = 0; i < SHA1_Message_Block_Size; ++i)
    context->Message_Block[i] = 0;
  context->Length_Low = 0;  /* and clear length */
  context->Length_High = 0;
  context->Computed = 1;
}

/*
 * SHA1PadMessage
 *
 * Description:
 *   According to the standard, the message must be padded to an
 *   even 512 bits. The first padding bit must be a '1'. The last
 *   64 bits represent the length of the original message. All bits
 *   in between should be 0. This helper function will pad the
 *   message according to those rules by filling the Message_Block
 *   array accordingly. When it returns, it can be assumed that the
 *   message digest has been computed.
 *
 * Parameters:
 *   context: [in/out]
 *     The context to pad
 *   Pad_Byte: [in]
 *     The last byte to add to the digest before the 0-padding
 *     and length. This will contain the last bits of the message
 *     followed by another single bit. If the message was an
 *     exact multiple of 8-bits long, Pad_Byte will be 0x80.
 *
 * Returns:
 *   Nothing.
 */
static void SHA1PadMessage(SHA1Context *context, uint8_t Pad_Byte)
{
  /*
   * Check to see if the current message block is too small to hold
   * the initial padding bits and length. If so, we will pad the
   * block, process it, and then continue padding into a second
   * block.
   */
  if (context->Message_Block_Index >= (SHA1_Message_Block_Size - 8)) {
    context->Message_Block[context->Message_Block_Index++] = Pad_Byte;
    while (context->Message_Block_Index < SHA1_Message_Block_Size)
      context->Message_Block[context->Message_Block_Index++] = 0;

    SHA1ProcessMessageBlock(context);
  } else
    context->Message_Block[context->Message_Block_Index++] = Pad_Byte;

  while (context->Message_Block_Index < (SHA1_Message_Block_Size - 8))
    context->Message_Block[context->Message_Block_Index++] = 0;

  /*
   * Store the message length as the last 8 octets
   */
  context->Message_Block[56] = (uint8_t) (context->Length_High >> 24);
  context->Message_Block[57] = (uint8_t) (context->Length_High >> 16);
  context->Message_Block[58] = (uint8_t) (context->Length_High >> 8);
  context->Message_Block[59] = (uint8_t) (context->Length_High);
  context->Message_Block[60] = (uint8_t) (context->Length_Low >> 24);
  context->Message_Block[61] = (uint8_t) (context->Length_Low >> 16);
  context->Message_Block[62] = (uint8_t) (context->Length_Low >> 8);
  context->Message_Block[63] = (uint8_t) (context->Length_Low);

  SHA1ProcessMessageBlock(context);
}

/*
 * SHA1ProcessMessageBlock
 *
 * Description:
 *   This helper function will process the next 512 bits of the
 *   message stored in the Message_Block array.
 *
 * Parameters:
 *   None.
 *
 * Returns:
 *   Nothing.
 *
 * Comments:
 *   Many of the variable names in this code, especially the
 *   single character names, were used because those were the
 *   names used in the publication.
 */
static void SHA1ProcessMessageBlock(SHA1Context *context)
{
  /* Constants defined in FIPS-180-2, section 4.2.1 */
  const uint32_t K[4] = {
      0x5A827999, 0x6ED9EBA1, 0x8F1BBCDC, 0xCA62C1D6
  };
  int        t;               /* Loop counter */
  uint32_t   temp;            /* Temporary word value */
  uint32_t   W[80];           /* Word sequence */
  uint32_t   A, B, C, D, E;   /* Word buffers */

  /*
   * Initialize the first 16 words in the array W
   */
  for (t = 0; t < 16; t++) {
    W[t]  = ((uint32_t)context->Message_Block[t * 4]) << 24;
    W[t] |= ((uint32_t)context->Message_Block[t * 4 + 1]) << 16;
    W[t] |= ((uint32_t)context->Message_Block[t * 4 + 2]) << 8;
    W[t] |= ((uint32_t)context->Message_Block[t * 4 + 3]);
  }
  for (t = 16; t < 80; t++)
    W[t] = SHA1_ROTL(1, W[t-3] ^ W[t-8] ^ W[t-14] ^ W[t-16]);

  A = context->Intermediate_Hash[0];
  B = context->Intermediate_Hash[1];
  C = context->Intermediate_Hash[2];
  D = context->Intermediate_Hash[3];
  E = context->Intermediate_Hash[4];

  for (t = 0; t < 20; t++) {
    temp = SHA1_ROTL(5,A) + SHA_Ch(B, C, D) + E + W[t] + K[0];
    E = D;
    D = C;
    C = SHA1_ROTL(30,B);
    B = A;
    A = temp;
  }

  for (t = 20; t < 40; t++) {
    temp = SHA1_ROTL(5,A) + SHA_Parity(B, C, D) + E + W[t] + K[1];
    E = D;
    D = C;
    C = SHA1_ROTL(30,B);
    B = A;
    A = temp;
  }

  for (t = 40; t < 60; t++) {
    temp = SHA1_ROTL(5,A) + SHA_Maj(B, C, D) + E + W[t] + K[2];
    E = D;
    D = C;
    C = SHA1_ROTL(30,B);
    B = A;
    A = temp;
  }

  for (t = 60; t < 80; t++) {
    temp = SHA1_ROTL(5,A) + SHA_Parity(B, C, D) + E + W[t] + K[3];
    E = D;
    D = C;
    C = SHA1_ROTL(30,B);
    B = A;
    A = temp;
  }

  context->Intermediate_Hash[0] += A;
  context->Intermediate_Hash[1] += B;
  context->Intermediate_Hash[2] += C;
  context->Intermediate_Hash[3] += D;
  context->Intermediate_Hash[4] += E;

  context->Message_Block_Index = 0;
}

